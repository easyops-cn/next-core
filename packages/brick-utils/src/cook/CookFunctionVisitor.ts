import {
  ArrowFunctionExpression,
  AssignmentExpression,
  BlockStatement,
  BreakStatement,
  CatchClause,
  ContinueStatement,
  DoWhileStatement,
  ExpressionStatement,
  ForInStatement,
  ForOfStatement,
  ForStatement,
  FunctionDeclaration,
  FunctionExpression,
  IfStatement,
  ReturnStatement,
  SwitchCase,
  SwitchStatement,
  ThrowStatement,
  TryStatement,
  UpdateExpression,
  VariableDeclaration,
  WhileStatement,
} from "@babel/types";
import { SimpleFunction } from "@next-core/brick-types";
import {
  CookVisitorState,
  ICookVisitor,
  VisitorCallback,
  VisitorFn,
} from "./interfaces";
import { CookVisitor } from "./CookVisitor";
import {
  assertIterable,
  isTerminated,
  spawnCookState,
  spawnCookStateOfBlock,
  lowerLevelSpawnCookStateOfBlock,
} from "./utils";

const ForOfStatementItemVisitor = (
  node: ForOfStatement | ForInStatement,
  blockState: CookVisitorState<void>,
  callback: VisitorCallback<CookVisitorState>,
  value: unknown
): void => {
  const leftState = spawnCookState(blockState, {
    assignment: {
      initializing: true,
      rightCooked: value,
    },
  });
  callback(node.left, leftState);
  callback(node.body, blockState);
};

const ForOfStatementVisitor: VisitorFn<CookVisitorState> = (
  node: ForOfStatement | ForInStatement,
  state,
  callback
) => {
  const blockState = spawnCookStateOfBlock(node, state, {
    controlFlow: {},
  });

  const rightState = spawnCookState<Iterable<unknown>>(blockState);
  callback(node.right, rightState);

  if (node.type === "ForOfStatement") {
    assertIterable(rightState.cooked, state.source, node.start, node.end);
    for (const value of rightState.cooked) {
      ForOfStatementItemVisitor(node, blockState, callback, value);
      blockState.controlFlow.continued = false;
      if (isTerminated(blockState)) {
        break;
      }
    }
  } else {
    for (const value in rightState.cooked) {
      ForOfStatementItemVisitor(node, blockState, callback, value);
      blockState.controlFlow.continued = false;
      if (isTerminated(blockState)) {
        break;
      }
    }
  }
};

const FunctionVisitor: VisitorFn<CookVisitorState> = (
  node: FunctionDeclaration | FunctionExpression | ArrowFunctionExpression,
  state,
  callback
) => {
  if (node.async || node.generator) {
    state.raiseError(
      SyntaxError,
      `${node.async ? "Async" : "Generator"} function is not allowed`,
      node
    );
  }

  if (
    node.type === "FunctionDeclaration" &&
    !(state.hoisting || state.isRoot)
  ) {
    return;
  }

  const fn: SimpleFunction = function (...args) {
    const { precookScope, blockState } = lowerLevelSpawnCookStateOfBlock(
      node,
      state,
      {
        isFunctionBody: true,
        returns: {
          returned: false,
        },
      }
    );

    if (node.type === "FunctionExpression" && node.id) {
      const topScope = blockState.scopeStack[blockState.scopeStack.length - 1];
      const ref = topScope.variables.get(node.id.name);
      ref.cooked = state.cooked;
      ref.initialized = true;
    }

    node.params.forEach((param, index) => {
      const variableInitValue =
        param.type === "RestElement" ? args.slice(index) : args[index];

      const paramState = spawnCookState(blockState, {
        assignment: {
          initializing: true,
          rightCooked: variableInitValue,
        },
      });

      callback(param, paramState);
    });

    for (const hoistedFn of precookScope.hoistedFunctions) {
      callback(
        hoistedFn,
        spawnCookState(blockState, {
          isFunctionBody: true,
          hoisting: true,
        })
      );
    }

    callback(node.body, blockState);

    return blockState.returns.cooked;
  };

  if (state.isRoot || node.type !== "FunctionDeclaration") {
    state.cooked = fn;
  }

  if (node.type === "FunctionDeclaration") {
    const topScope = state.scopeStack[state.scopeStack.length - 1];
    const ref = topScope.get(node.id.name);
    ref.cooked = fn;
    ref.initialized = true;
  }
};

export const CookFunctionVisitor = Object.freeze({
  ...CookVisitor,
  ArrowFunctionExpression: FunctionVisitor,
  AssignmentExpression(node: AssignmentExpression, state, callback) {
    const rightState = spawnCookState(state);
    callback(node.right, rightState);

    const leftState = spawnCookState(state, {
      assignment: {
        operator: node.operator,
        rightCooked: rightState.cooked,
      },
    });
    callback(node.left, leftState);

    state.cooked = leftState.cooked;
  },
  BlockStatement(node: BlockStatement, state, callback) {
    const { precookScope, blockState } = lowerLevelSpawnCookStateOfBlock(
      node,
      state
    );

    if (precookScope) {
      for (const hoistedFn of precookScope.hoistedFunctions) {
        callback(
          hoistedFn,
          spawnCookState(blockState, {
            hoisting: true,
          })
        );
      }
    }

    for (const statement of node.body) {
      callback(statement, blockState);
      if (isTerminated(blockState)) {
        break;
      }
    }
  },
  BreakStatement(node: BreakStatement, state) {
    // istanbul ignore if
    if (node.label) {
      state.raiseError(
        SyntaxError,
        "Labeled break statement is not allowed",
        node
      );
    }
    state.controlFlow.broken = true;
  },
  CatchClause(node: CatchClause, state, callback) {
    const blockState = spawnCookStateOfBlock(node, state);

    const paramState = spawnCookState(blockState, {
      assignment: {
        initializing: true,
        rightCooked: state.caughtError,
      },
    });
    callback(node.param, paramState);
    callback(node.body, spawnCookState(blockState));
  },
  ContinueStatement(node: ContinueStatement, state) {
    // istanbul ignore if
    if (node.label) {
      state.raiseError(
        SyntaxError,
        "Labeled continue statement is not allowed",
        node
      );
    }
    state.controlFlow.continued = true;
  },
  EmptyStatement() {
    // Do nothing.
  },
  ExpressionStatement(node: ExpressionStatement, state, callback) {
    callback(node.expression, spawnCookState(state));
  },
  ForInStatement: ForOfStatementVisitor,
  ForOfStatement: ForOfStatementVisitor,
  ForStatement(node: ForStatement, state, callback) {
    const blockState = spawnCookStateOfBlock(node, state, {
      controlFlow: {},
    });
    if (node.init) {
      callback(node.init, spawnCookState(blockState));
    }
    for (
      let testState: CookVisitorState;
      node.test
        ? (callback(node.test, (testState = spawnCookState(blockState))),
          testState.cooked)
        : true;
      node.update && callback(node.update, spawnCookState(blockState))
    ) {
      callback(node.body, spawnCookState(blockState));
      blockState.controlFlow.continued = false;
      if (isTerminated(blockState)) {
        break;
      }
    }
  },
  FunctionDeclaration: FunctionVisitor,
  FunctionExpression: FunctionVisitor,
  IfStatement(node: IfStatement, state, callback) {
    const testState = spawnCookState(state);
    callback(node.test, testState);
    if (testState.cooked) {
      callback(node.consequent, spawnCookState(state));
    } else if (node.alternate) {
      callback(node.alternate, spawnCookState(state));
    }
  },
  ReturnStatement(node: ReturnStatement, state, callback) {
    const argumentState = spawnCookState(state);
    if (node.argument) {
      callback(node.argument, argumentState);
    }
    state.returns.returned = true;
    state.returns.cooked = argumentState.cooked;
  },
  SwitchCase(node: SwitchCase, state, callback) {
    if (!state.controlFlow.switchTested && node.test) {
      const testState = spawnCookState(state);
      callback(node.test, testState);
      state.controlFlow.switchTested =
        testState.cooked === state.controlFlow.switchDiscriminantCooked;
    }
    if (state.controlFlow.switchTested || !node.test) {
      for (const statement of node.consequent) {
        callback(statement, state);
        if (isTerminated(state)) {
          state.controlFlow.switchTested = false;
          break;
        }
      }
    }
  },
  SwitchStatement(node: SwitchStatement, state, callback) {
    const discriminantState = spawnCookState(state);
    callback(node.discriminant, discriminantState);

    const { precookScope, blockState } = lowerLevelSpawnCookStateOfBlock(
      node,
      state,
      {
        controlFlow: {
          switchDiscriminantCooked: discriminantState.cooked,
        },
      }
    );

    for (const hoistedFn of precookScope.hoistedFunctions) {
      callback(
        hoistedFn,
        spawnCookState(blockState, {
          hoisting: true,
        })
      );
    }

    for (const switchCase of node.cases) {
      callback(switchCase, blockState);
      if (isTerminated(blockState)) {
        break;
      }
    }
  },
  ThrowStatement(node: ThrowStatement, state, callback) {
    const throwState = spawnCookState(state);
    callback(node.argument, throwState);
    throw throwState.cooked;
  },
  TryStatement(node: TryStatement, state, callback) {
    try {
      callback(node.block, spawnCookState(state));
    } catch (error) {
      if (node.handler) {
        callback(
          node.handler,
          spawnCookState(state, {
            caughtError: error,
          })
        );
      } else {
        throw error;
      }
    } finally {
      if (node.finalizer) {
        callback(node.finalizer, spawnCookState(state));
      }
    }
  },
  UpdateExpression(node: UpdateExpression, state, callback) {
    const argumentState = spawnCookState(state, {
      update: {
        operator: node.operator,
        prefix: node.prefix,
      },
    });
    callback(node.argument, argumentState);
    state.cooked = argumentState.cooked;
  },
  VariableDeclaration(node: VariableDeclaration, state, callback) {
    if (node.kind === "var" && state.rules.noVar) {
      state.raiseError(
        SyntaxError,
        "Var declaration is not recommended, use `let` or `const` instead",
        node
      );
    }
    for (const declaration of node.declarations) {
      let initCooked;
      let hasInit = false;
      if (state.assignment) {
        initCooked = state.assignment.rightCooked;
        hasInit = true;
      } else if (declaration.init) {
        const initState = spawnCookState(state);
        callback(declaration.init, initState);
        initCooked = initState.cooked;
        hasInit = true;
      }
      if (node.kind !== "var" || hasInit) {
        const idState = spawnCookState(state, {
          assignment: {
            initializing: true,
            rightCooked: initCooked,
          },
        });
        callback(declaration.id, idState);
      }
    }
  },
  WhileStatement(node: WhileStatement, state, callback) {
    let testState: CookVisitorState;
    const blockState = spawnCookState<void>(state, {
      controlFlow: {},
    });
    while (
      (callback(node.test, (testState = spawnCookState(state))),
      testState.cooked)
    ) {
      callback(node.body, blockState);
      blockState.controlFlow.continued = false;
      if (isTerminated(blockState)) {
        break;
      }
    }
  },
  DoWhileStatement(node: DoWhileStatement, state, callback) {
    let testState: CookVisitorState;
    const blockState = spawnCookState<void>(state, {
      controlFlow: {},
    });
    do {
      callback(node.body, blockState);
      blockState.controlFlow.continued = false;
      if (isTerminated(blockState)) {
        break;
      }
    } while (
      (callback(node.test, (testState = spawnCookState(state))),
      testState.cooked)
    );
  },
} as ICookVisitor);
