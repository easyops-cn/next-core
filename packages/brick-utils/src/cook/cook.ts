import { Node } from "@babel/types";
import { raiseErrorFactory, walkFactory } from "./utils";
import { CookVisitor } from "./CookVisitor";
import { CookVisitorState, PrecookResult } from "./interfaces";
import { supply } from "./supply";

export function cook<T = unknown>(
  precooked: PrecookResult,
  globalVariables?: Record<string, unknown>
): T {
  const raiseError = raiseErrorFactory(precooked.source);
  const state: CookVisitorState<T> = {
    source: precooked.source,
    raiseError,
    scopeMapByNode: precooked.scopeMapByNode,
    scopeStack: [supply(precooked.attemptToVisitGlobals, globalVariables)],
    rules: {},
  };
  walkFactory(CookVisitor, (node: Node) => {
    raiseError(SyntaxError, `Unsupported node type \`${node.type}\``, node);
  })(precooked.expression, state);
  return state.cooked;
}
