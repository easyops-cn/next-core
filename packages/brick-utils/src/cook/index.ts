export { cook } from "./cook";
export { precook } from "./precook";
export { cookFunction } from "./cookFunction";
export { precookFunction } from "./precookFunction";
export * from "./preevaluate";
export { PrecookVisitor } from "./PrecookVisitor";
export { PrecookFunctionVisitor } from "./PrecookFunctionVisitor";
