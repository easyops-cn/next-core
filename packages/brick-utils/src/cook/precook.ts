import { Node } from "@babel/types";
import { parseExpression } from "@babel/parser";
import { walkFactory } from "./utils";
import { PrecookVisitor } from "./PrecookVisitor";
import {
  PrecookVisitorState,
  PrecookResult,
  PrecookOptions,
} from "./interfaces";

export function precook(
  source: string,
  options?: PrecookOptions
): PrecookResult {
  const state: PrecookVisitorState = {
    scopeStack: [],
    attemptToVisitGlobals: new Set(),
    scopeMapByNode: new WeakMap(),
  };
  const expression = parseExpression(source, {
    plugins: ["estree", ["pipelineOperator", { proposal: "minimal" }]],
  });

  // const attemptToVisitMembers = new Map<string, Set<string>>();
  walkFactory(
    options?.visitors
      ? { ...PrecookVisitor, ...options.visitors }
      : PrecookVisitor,
    (node: Node) => {
      // eslint-disable-next-line no-console
      console.warn(
        `Unsupported node type \`${node.type}\`: \`${source.substring(
          node.start,
          node.end
        )}\``
      );
    }
  )(expression, state);

  return {
    source,
    expression,
    attemptToVisitGlobals: state.attemptToVisitGlobals,
    scopeMapByNode: state.scopeMapByNode,
  };
}
