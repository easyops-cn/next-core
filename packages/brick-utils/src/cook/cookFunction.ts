import { Node } from "@babel/types";
import { SimpleFunction } from "@next-core/brick-types";
import { raiseErrorFactory, walkFactory } from "./utils";
import { CookFunctionVisitor } from "./CookFunctionVisitor";
import {
  CookFunctionOptions,
  CookVisitorState,
  PrecookFunctionResult,
} from "./interfaces";
import { supply } from "./supply";

export function cookFunction<T extends SimpleFunction>(
  precooked: PrecookFunctionResult,
  { rules = {}, globalVariables }: CookFunctionOptions = {}
): T {
  const raiseError = raiseErrorFactory(precooked.source);
  const state: CookVisitorState<T> = {
    source: precooked.source,
    raiseError,
    scopeMapByNode: precooked.scopeMapByNode,
    scopeStack: [supply(precooked.attemptToVisitGlobals, globalVariables)],
    isRoot: true,
    rules,
  };
  walkFactory(CookFunctionVisitor, (node: Node) => {
    raiseError(SyntaxError, `Unsupported node type \`${node.type}\``, node);
  })(precooked.function, state);
  return state.cooked;
}
