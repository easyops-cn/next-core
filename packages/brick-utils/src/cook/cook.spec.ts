import { install, InstalledClock } from "lolex";
import { cook } from "./cook";
import { precook } from "./precook";

jest.spyOn(console, "warn").mockImplementation(() => void 0);

jest.mock("../placeholder/pipes", () => ({
  PipeRegistry: new Map([
    ["string", (v: unknown) => (v == null ? "" : String(v))],
  ]),
}));

describe("cook", () => {
  let clock: InstalledClock;
  beforeEach(() => {
    clock = install({ now: +new Date("2020-03-25 17:37:00") });
  });
  afterEach(() => {
    clock.uninstall();
  });

  const getGlobalVariables = (): Record<string, unknown> => ({
    DATA: {
      for: "good",
      null: null,
      undefined: undefined,
      true: true,
      false: false,
      number5: 5,
      objectA: {
        onlyInA: 1,
        bothInAB: 2,
      },
      objectB: {
        onlyInB: 3,
        bothInAB: 4,
      },
      q: "a&b",
      redirect: "/r/s?t=u&v=w",
      path: "x/y.zip",
    },
    APP: {
      homepage: "/hello/world",
    },
  });

  it.each<[string, unknown]>([
    ["'good'", "good"],
    ["1", 1],
    ["null", null],
    ["undefined", undefined],
    ["true", true],
    ["NaN", NaN],
    ["/bc/.test('abcd')", true],
    ["/bc/.test('dcba')", false],
    ["isNaN(NaN)", true],
    ["isNaN({})", true],
    ["isNaN(1)", false],
    ["Array.isArray([])", true],
    ["Array.isArray({})", false],
    ["Object.keys(DATA.objectA)", ["onlyInA", "bothInAB"]],
    ["String([1, 2])", "1,2"],
    ["Boolean(1)", true],
    ["Boolean(0)", false],
    ["Number(true)", 1],
    ["Number.isNaN({})", false],
    ["JSON.parse('[1]')", [1]],
    ["isFinite(5)", true],
    ["isFinite(Infinity)", false],
    ["parseFloat('5.3good')", 5.3],
    ["parseInt('5.3good')", 5],
    ["Math.max(1, 2)", 2],
    ["PIPES.string(null)", ""],
    ["location.href", "http://localhost/"],
    ["DATA.for", "good"],
    ["DATA['for']", "good"],
    ["DATA.other", undefined],
    ["{}", {}],
    [
      "{ quality: DATA.for, [DATA.true]: 'story' }",
      {
        quality: "good",
        true: "story",
      },
    ],
    ["[]", []],
    ["[1, DATA.number5]", [1, 5]],
    [
      // `ArrowFunctionExpression` mixed `CallExpression`
      "(a => a.b)({b: 'c'})",
      "c",
    ],
    [
      // `MemberExpression`
      "DATA.number5.toFixed(1)",
      "5.0",
    ],
    [
      // `OptionalMemberExpression`
      "DATA.number5?.toFixed(1)",
      "5.0",
    ],
    ["DATA.null || 'oops'", "oops"],
    ["DATA.for || 'oops'", "good"],
    ["DATA.null && 'oops'", null],
    ["DATA.for && 'oops'", "oops"],
    ["DATA.for ?? 'oops'", "good"],
    ["DATA.false ?? 'oops'", false],
    ["DATA.null ?? 'oops'", "oops"],
    ["DATA.undefined ?? 'oops'", "oops"],
    ["DATA.for?.length", 4],
    ["String?.(null)", "null"],
    ["DATA.notExisted?.length", undefined],
    ["DATA.notExisted?.length?.oops", undefined],
    ["DATA.notExisted?.()", undefined],
    ["DATA.notExisted?.()?.()", undefined],
    ["DATA.notExisted?.()?.length", undefined],
    ["DATA.notExisted?.length?.()", undefined],
    ["DATA.notExisted?.length.oops", undefined],
    ["DATA.notExisted?.length()", undefined],
    ["DATA.notExisted?.()()", undefined],
    ["(DATA.notExisted)?.length", undefined],
    ["(DATA.notExisted?.length)?.oops", undefined],
    ["!DATA.null", true],
    ["+DATA.true", 1],
    ["-DATA.true", -1],
    ["typeof DATA.for", "string"],
    ["typeof DATA.unknown", "undefined"],
    ["typeof unknown", "undefined"],
    ["void DATA.for", undefined],
    ["DATA.number5 + 1", 6],
    ["DATA.number5 - 1", 4],
    ["DATA.number5 / 2", 2.5],
    ["DATA.number5 % 2", 1],
    ["DATA.number5 * 2", 10],
    ["DATA.number5 ** 2", 25],
    ["DATA.number5 == '5'", true],
    ["DATA.number5 == 4", false],
    ["DATA.number5 === 5", true],
    ["DATA.number5 === '5'", false],
    ["DATA.number5 != '5'", false],
    ["DATA.number5 != 4", true],
    ["DATA.number5 !== 5", false],
    ["DATA.number5 !== '5'", true],
    ["DATA.number5 > 4", true],
    ["DATA.number5 > 5", false],
    ["DATA.number5 < 6", true],
    ["DATA.number5 < 5", false],
    ["DATA.number5 >= 4", true],
    ["DATA.number5 >= 5", true],
    ["DATA.number5 >= 6", false],
    ["DATA.number5 <= 6", true],
    ["DATA.number5 <= 5", true],
    ["DATA.number5 <= 4", false],
    ["DATA.for ? 'yep': 'oops'", "yep"],
    ["DATA.null ? 'yep': 'oops'", "oops"],
    [
      // `SequenceExpression`
      "DATA.for, DATA.number5",
      5,
    ],
    [
      "`${null},${undefined},${true},${false},${{}},${[]},${[1,2]}${5},${NaN}`",
      `${null},${undefined},${true},${false},${{}},${[]},${[1, 2]}${5},${NaN}`,
    ],
    ["_.get(DATA, 'for')", "good"],
    ["moment().format()", "2020-03-25T17:37:00+08:00"],
    ["moment('not a real date').isValid()", false],
    [
      "moment('12-25-1995', 'MM-DD-YYYY').format(moment.HTML5_FMT.DATETIME_LOCAL)",
      "1995-12-25T00:00",
    ],
    ["[1,2,3].slice(1)", [2, 3]],
    ["[1, ...DATA.for, 2]", [1, "g", "o", "o", "d", 2]],
    // eslint-disable-next-line no-sparse-arrays
    ["[1, , 2]", [1, , 2]],
    ["[-1].concat(0, ...[1, 2], 3)", [-1, 0, 1, 2, 3]],
    ["[-1]?.concat(0, ...[1, 2], 3)", [-1, 0, 1, 2, 3]],
    [
      "{...DATA.objectA, ...DATA.objectB}",
      {
        onlyInA: 1,
        bothInAB: 4,
        onlyInB: 3,
      },
    ],
    ["{...null, ...undefined}", {}],
    ["[1, undefined, null].map((i = 5) => i)", [1, 5, null]],
    ["[1, undefined].map((i = DATA.number5) => i)", [1, 5]],
    [
      // `j` is not defined, but not evaluated either.
      "[1, 2].map((i = j + 1) => i)",
      [1, 2],
    ],
    [
      // `j` is not initialized, but not evaluated either.
      "[1, 2].map((i = j + 1, j) => i)",
      [1, 2],
    ],
    ["[1, 2].map(i => ((i, j = i + 1) => i + j)(i))", [3, 5]],
    [
      "[1, 2].map((...args) => args)",
      [
        [1, 0, [1, 2]],
        [2, 1, [1, 2]],
      ],
    ],
    [
      "[1, 2].map((i, ...args) => args)",
      [
        [0, [1, 2]],
        [1, [1, 2]],
      ],
    ],
    // `ArrayPattern`
    ["[[1, 2]].map(([a, b]) => a + b)", [3]],
    // `ArrayPattern` with `RestElement`
    ["[[1, 2, 3]].map(([a, ...b]) => a + b.length)", [3]],
    // Nested `ArrayPattern`
    ["[[1, [2, 3]]].map(([a, [b, c]]) => a + b + c)", [6]],
    // `ArrayPattern` with `AssignmentPattern`
    ["[[1]].map(([a, [b, c] = [2, 3]]) => a + b + c)", [6]],
    // `ArrayPattern` with Nested `AssignmentPattern`
    ["[[1]].map(([a, [b, c = 3] = [2]]) => a + b + c)", [6]],
    // `ArrayPattern` with parameter scope
    ["[[1]].map(([a, [b, c = b] = [2]]) => a + b + c)", [5]],
    // `ObjectPattern`
    ["[{a: 1, b: 2}].map(({a, b}) => a + b)", [3]],
    // `ObjectPattern` with `RestElement`
    ["[{a: 1, b: 2, c: 3}].map(({a, ...b}) => a + b.b + b.c)", [6]],
    // Nested `ObjectPattern`
    ["[{a: 1, b: { d: 2 }}].map(({a, b: { d: c }}) => a + c)", [3]],
    // `ObjectPattern` with `AssignmentPattern` and `RestElement`
    ["[undefined].map(({a, ...b}={}) => a + b)", ["undefined[object Object]"]],
    // `ObjectPattern` with a computed key
    ["[{'a.b': 1}].map(({'a.b': c}) => c)", [1]],
    // Pipeline operator.
    ["DATA.number5 |> PIPES.string", "5"],
    // Sequential pipeline operators with an arrow function.
    ["DATA.number5 |> (_ => _ + 1) |> PIPES.string", "6"],
    // Reuse arrow functions.
    ["(fn => fn(2)+fn())((a=1)=>a)", 3],
    // Nested arrow functions
    ["((a)=>(b)=>a+b)(1)(2)", 3],
    ["new Set([1, 2, 3])", new Set([1, 2, 3])],
    ["new Array(1, ...[2, 3])", [1, 2, 3]],
    [
      "String(new URLSearchParams({q: 'hello,world', age: 18}))",
      "q=hello%2Cworld&age=18",
    ],
    // Tagged template.
    ["((s,...k) => `${s.join('-')}:${k.join(',')}`)`a${1}b${2}c`", "a-b-c:1,2"],
    [
      "TAG_URL`${APP.homepage}/list?q=${DATA.q}&redirect=${DATA.redirect}`",
      "/hello/world/list?q=a%26b&redirect=/r/s%3Ft%3Du%26v%3Dw",
    ],
    ["SAFE_TAG_URL`file/${DATA.path}?q=${DATA.q}`", "file/x%2Fy.zip?q=a%26b"],
    ["btoa('hello')", "aGVsbG8="],
    ["atob('aGVsbG8=')", "hello"],
  ])("cook(precook(%j), {...}) should return %j", (input, cooked) => {
    expect(cook(precook(input), getGlobalVariables())).toEqual(cooked);
  });

  it.each<string>([
    "this.bad",
    "() => {}",
    // `_.reverse` is not supplied.
    "_.reverse([0,1,2])",
    "Object.assign(DATA, { override: true })",
    // When `EVENT` is not supplied.
    "EVENT.detail",
    "delete DATA.for",
    "DATA.number5 ^ 1",
    "DATA.number5()",
    "DATA?.()",
    "DATA?.number5()",
    "DATA?.number5.notExisted.oops",
    "DATA.number5?.toFixed.oo.ps",
    "DATA.number5.toFixed?.().oops()",
    "(DATA.notExisted?.length).oops",
    "[...DATA]",
    "[...null]",
    "[...undefined]",
    // `j` is not defined, and will be evaluated.
    "[undefined].map((i = j + 1) => i)",
    // `j` is not initialized, and will be evaluated.
    "[undefined].map((i = j + 1, j) => i)",
    // Parameters affect the scopes.
    "[undefined].map((i = DATA.number5, DATA) => i)",
    // `ObjectPattern` meet nil
    "[undefined].map(({}) => 1)",
    "[undefined].map(({a}) => a)",
    "[undefined].map(({...a}) => a)",
    "async () => null",
    "location.assign('/')",
    "moment.updateLocale('en', {})",
    // `ArrayPattern` meets non-iterable.
    "[[1]].map(([a, [b]]) => a + b)",
    "[1, 2].map(([, a]) => a)",
    "/bc\\u{/u.test('dcba')",
    "/bc\\u{13}/u.test('dcba')",
    "({}).constructor.assign",
    "((a,b)=>a[b])(()=>1, 'constructor')",
    "((a,b)=>a[b])(()=>1, 'constructor')('console.log(`yo`)')()",
    "((a,b)=>a[b])(()=>1, 'constructor').bind(null)('console.log(`yo`)')()",
    "_.get(()=>1, 'constructor.prototype')",
    "DATA |> DATA.number5",
    "new Boolean()",
    "new (a = function () {})()",
    "(Set => new Set())(() => null)",
    "1`a${1}b`",
    "c`a${1}b`",
    // Reuse arrow functions.
    "(fn => fn(2,1)+fn())((a=b,b)=>a)",
    "typeof unknown.any",
    // Todo(steve)
    // "_.wrap(_.method('constructor.assign',{a:1},{b:2}),(func,...a) => func(...a))({})"
  ])("cook(precook(%j), {...}) should throw", (input) => {
    expect(() =>
      cook(precook(input), getGlobalVariables())
    ).toThrowErrorMatchingSnapshot();
  });
});
