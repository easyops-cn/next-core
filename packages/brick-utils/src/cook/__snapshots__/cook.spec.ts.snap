// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cook cook(precook("((a,b)=>a[b])(()=>1, 'constructor')"), {...}) should throw 1`] = `"Cannot access reserved objects such as \`Function\`."`;

exports[`cook cook(precook("((a,b)=>a[b])(()=>1, 'constructor')('console.log(\`yo\`)')()"), {...}) should throw 1`] = `"Cannot access reserved objects such as \`Function\`."`;

exports[`cook cook(precook("((a,b)=>a[b])(()=>1, 'constructor').bind(null)('console.log(\`yo\`)')()"), {...}) should throw 1`] = `"Cannot access reserved objects such as \`Function\`."`;

exports[`cook cook(precook("() => {}"), {...}) should throw 1`] = `"Only an \`Expression\` is allowed in \`ArrowFunctionExpression\`'s body: \`() => {}\`"`;

exports[`cook cook(precook("({}).constructor.assign"), {...}) should throw 1`] = `"Cannot access reserved objects such as \`Function\`."`;

exports[`cook cook(precook("(DATA.notExisted?.length).oops"), {...}) should throw 1`] = `"Cannot read property 'oops' of undefined"`;

exports[`cook cook(precook("(Set => new Set())(() => null)"), {...}) should throw 1`] = `"Unsupported non-global constructor \`Set\`: \`new Set()\`"`;

exports[`cook cook(precook("(fn => fn(2,1)+fn())((a=b,b)=>a)"), {...}) should throw 1`] = `"Cannot access 'b' before initialization"`;

exports[`cook cook(precook("/bc\\\\u{/u.test('dcba')"), {...}) should throw 1`] = `"Invalid regular expression: /bc\\\\u{/u"`;

exports[`cook cook(precook("/bc\\\\u{13}/u.test('dcba')"), {...}) should throw 1`] = `"Unsupported unicode flag in regular expression: /bc\\\\u{13}/u"`;

exports[`cook cook(precook("[...DATA]"), {...}) should throw 1`] = `"object is not iterable: \`...DATA\`"`;

exports[`cook cook(precook("[...null]"), {...}) should throw 1`] = `"object is not iterable: \`...null\`"`;

exports[`cook cook(precook("[...undefined]"), {...}) should throw 1`] = `"undefined is not iterable: \`...undefined\`"`;

exports[`cook cook(precook("[[1]].map(([a, [b]]) => a + b)"), {...}) should throw 1`] = `"undefined is not iterable: \`[b]\`"`;

exports[`cook cook(precook("[1, 2].map(([, a]) => a)"), {...}) should throw 1`] = `"number is not iterable: \`[, a]\`"`;

exports[`cook cook(precook("[undefined].map(({...a}) => a)"), {...}) should throw 1`] = `"Cannot destructure undefined"`;

exports[`cook cook(precook("[undefined].map(({}) => 1)"), {...}) should throw 1`] = `"Cannot destructure undefined"`;

exports[`cook cook(precook("[undefined].map(({a}) => a)"), {...}) should throw 1`] = `"Cannot destructure undefined"`;

exports[`cook cook(precook("[undefined].map((i = DATA.number5, DATA) => i)"), {...}) should throw 1`] = `"Cannot access 'DATA' before initialization"`;

exports[`cook cook(precook("[undefined].map((i = j + 1) => i)"), {...}) should throw 1`] = `"j is not defined"`;

exports[`cook cook(precook("[undefined].map((i = j + 1, j) => i)"), {...}) should throw 1`] = `"Cannot access 'j' before initialization"`;

exports[`cook cook(precook("_.get(()=>1, 'constructor.prototype')"), {...}) should throw 1`] = `"Cannot access reserved objects such as \`Function\`."`;

exports[`cook cook(precook("_.reverse([0,1,2])"), {...}) should throw 1`] = `"_.reverse is not a function"`;

exports[`cook cook(precook("1\`a\${1}b\`"), {...}) should throw 1`] = `"1 is not a function"`;

exports[`cook cook(precook("DATA |> DATA.number5"), {...}) should throw 1`] = `"DATA.number5 is not a function"`;

exports[`cook cook(precook("DATA.number5 ^ 1"), {...}) should throw 1`] = `"Unsupported binary operator \`^\`: \`DATA.number5 ^ 1\`"`;

exports[`cook cook(precook("DATA.number5()"), {...}) should throw 1`] = `"DATA.number5 is not a function"`;

exports[`cook cook(precook("DATA.number5.toFixed?.().oops()"), {...}) should throw 1`] = `"DATA.number5.toFixed?.().oops is not a function"`;

exports[`cook cook(precook("DATA.number5?.toFixed.oo.ps"), {...}) should throw 1`] = `"Cannot read property 'ps' of undefined"`;

exports[`cook cook(precook("DATA?.()"), {...}) should throw 1`] = `"DATA is not a function"`;

exports[`cook cook(precook("DATA?.number5()"), {...}) should throw 1`] = `"DATA?.number5 is not a function"`;

exports[`cook cook(precook("DATA?.number5.notExisted.oops"), {...}) should throw 1`] = `"Cannot read property 'oops' of undefined"`;

exports[`cook cook(precook("EVENT.detail"), {...}) should throw 1`] = `"EVENT is not defined"`;

exports[`cook cook(precook("Object.assign(DATA, { override: true })"), {...}) should throw 1`] = `"Object.assign is not a function"`;

exports[`cook cook(precook("async () => null"), {...}) should throw 1`] = `"Async function is not allowed: \`async () => null\`"`;

exports[`cook cook(precook("c\`a\${1}b\`"), {...}) should throw 1`] = `"c is not defined"`;

exports[`cook cook(precook("delete DATA.for"), {...}) should throw 1`] = `"Unsupported unary operator \`delete\`: \`delete DATA.for\`"`;

exports[`cook cook(precook("location.assign('/')"), {...}) should throw 1`] = `"location.assign is not a function"`;

exports[`cook cook(precook("moment.updateLocale('en', {})"), {...}) should throw 1`] = `"moment.updateLocale is not a function"`;

exports[`cook cook(precook("new (a = function () {})()"), {...}) should throw 1`] = `"Unsupported new expression: \`new (a = function () {})()\`"`;

exports[`cook cook(precook("new Boolean()"), {...}) should throw 1`] = `"Unsupported constructor \`Boolean\`: \`new Boolean()\`"`;

exports[`cook cook(precook("this.bad"), {...}) should throw 1`] = `"Unsupported node type \`ThisExpression\`: \`this\`"`;

exports[`cook cook(precook("typeof unknown.any"), {...}) should throw 1`] = `"unknown is not defined"`;
